q2_composition/_ancombc.py

```diff
@@ -0,0 +1,159 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2022, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+import subprocess
+import tempfile
+import pandas as pd
+import os
+import formulaic
+
+import qiime2
+
+from q2_stats._format import DataLoafPackageDirFmt
+
+
+def run_commands(cmds, verbose=True):
+    if verbose:
+        print('Running external command line application(s). This may print'
+              ' messages to stdout and/or stderr.')
+        print('The command(s) being run are below. These commands cannot be'
+              ' manually re-run as they will depend on temporary files that no'
+              ' longer exist.')
+    for cmd in cmds:
+        if verbose:
+            print('\nCommand:', end=' ')
+            print(' '.join(cmd), end='\n\n')
+        subprocess.run(cmd, check=True)
+
+
+def ancombc(table: pd.DataFrame, metadata: qiime2.Metadata, formula: str,
+            p_adj_method: str = 'holm', prv_cut: float = 0.1, lib_cut: int = 0,
+            level_ordering: str = None, neg_lb: bool = False,
+            tol: float = 1e-05, max_iter: int = 100, conserve: bool = False,
+            alpha: float = 0.05) -> DataLoafPackageDirFmt:
+
+    return _ancombc(
+        table=table,
+        metadata=metadata,
+        formula=formula,
+        p_adj_method=p_adj_method,
+        prv_cut=prv_cut,
+        lib_cut=lib_cut,
+        level_ordering=level_ordering,
+        neg_lb=neg_lb,
+        tol=tol,
+        max_iter=max_iter,
+        conserve=conserve,
+        alpha=alpha,
+    )
+
+
+# utility functions for formula parsing and column validation
+def _parse_terms(formula):
+    parse = formulaic.parser.parser.DefaultFormulaParser(
+        include_intercept=False)
+    terms = parse.get_ast(formula=formula).flatten()
+    formula_terms = _leaf_collector(terms)
+    return formula_terms
+
+
+def _leaf_collector(term):
+    if isinstance(term, formulaic.parser.types.Token):
+        return [str(term)]
+
+    if type(term) is not list:
+        return []
+
+    return _leaf_collector(term[1]) + _leaf_collector(term[2])
+
+
+def _column_validation(value, parameter, metadata):
```

Comment:
This function is a bit confusing to me. We don't use the `parameter` parameter at all outside of the ValueError, so what is its purpose? It seems like it really gets used outside of the function and we're handing it in here to describe some external state, but it's a bit confusing because you need to look at the call site to really have an idea of what's happening. Is it possible to refactor things to make it a bit more consistent if that makes sense? It looks like that would be tricky though. Additionally (and I think this is pretty subjective) I would probably make `metadata` the first parameter in the parameter list to be consistent with our other functions where inputs and metadata come before parameters.

Reply:
Hey @Oddant1 - yeah, it's a bit confusing. The `parameter` parameter is just used for the ValueError message to provide a more concise message to the user (i.e. value provided in the 'formula' parameter was not found... etc). I originally had this written separately for each parameter it was being used for, but added this level of abstraction to DRY it up a bit. Since we've removed the `group` parameter though, this is only being used for `formula` and `level_ordering`. I'm not opposed to re-writing it separately for each of those params, but will see what @ebolyen thinks is the best way to structure it.

I can definitely change the ordering for the args in it to include metadata first - I think that makes sense to have the order be consistent with the rest of the functions.

Reply:
I think this makes sense, and the utility at least avoids writing essentially the same error twice. I can see how the removal of `group` made it less obviously useful, but I think it's still worth it.